<a href="https://colab.research.google.com/github/Andrii-Oliinyk-2726/Data_Science_HW7/blob/main/Hw7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
pip install surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 6.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=3156231 sha256=9db93892397719f9119e6e1561c74d837ae90b36aae520f373c32e8900d81175
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


In [9]:
import pandas as pd
from surprise import Dataset, Reader
from surprise import SVDpp, SVD, NMF
from surprise.model_selection import train_test_split, cross_validate, GridSearchCV
from surprise import accuracy

# Load the movielens-100k dataset (download it if needed).
data = Dataset.load_builtin('ml-100k')


# **SVD**

In [18]:
algo_svd = SVD()
result_svd = cross_validate(algo_svd, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)
SVD_result = pd.DataFrame.from_dict(result_svd).mean(axis=0)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9414  0.9314  0.9364  0.9382  0.9366  0.9368  0.0032  
MAE (testset)     0.7409  0.7346  0.7365  0.7409  0.7380  0.7382  0.0025  
Fit time          2.16    1.43    1.42    1.41    1.60    1.60    0.29    
Test time         0.29    0.13    0.16    0.12    0.12    0.16    0.06    


# **SVD++**

In [19]:
algo_svdpp = SVDpp()
result_svdpp = cross_validate(algo_svdpp, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)
SVDpp_result = pd.DataFrame.from_dict(result_svdpp).mean(axis=0)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9288  0.9113  0.9170  0.9189  0.9161  0.9184  0.0058  
MAE (testset)     0.7266  0.7163  0.7182  0.7206  0.7208  0.7205  0.0035  
Fit time          27.34   28.79   28.25   27.20   28.58   28.03   0.65    
Test time         7.29    4.87    5.31    5.64    5.04    5.63    0.87    


# **NMF**

In [20]:
algo_nmf = SVDpp()
result_nmf = cross_validate(algo_nmf, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)
NMF_result = pd.DataFrame.from_dict(result_nmf).mean(axis=0)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9234  0.9153  0.9231  0.9141  0.9199  0.9192  0.0039  
MAE (testset)     0.7222  0.7174  0.7254  0.7151  0.7212  0.7203  0.0036  
Fit time          28.43   26.95   27.46   27.30   28.60   27.75   0.65    
Test time         4.51    5.64    4.71    5.82    4.74    5.08    0.54    


# **Conclusions**

In [21]:
Final_result = pd.DataFrame(columns=['SVD', 'SVDpp', 'NMF'])

Final_result['SVD'] = SVD_result
Final_result['SVDpp'] = SVDpp_result
Final_result['NMF'] = NMF_result

Final_result

SVD      SVDpp        NMF
test_rmse  0.936795   0.918425   0.919155
test_mae   0.738199   0.720514   0.720262
fit_time   1.602885  28.030748  27.748391
test_time  0.164744   5.632779   5.084527

SVD (матрична факторизація) має середнє значення RMSE приблизно 0.936795 та середнє значення MAE близько 0.738199, час обчислення - найменший. Доцільно використовувати для обчислення великих датасетів, при цьому враховувати, що точність результату може бути не найкращою.

SVD++ (розширена матрична факторизація) показує кращі результати з середнім значенням RMSE приблизно 0.918425 та середнім значенням MAE близько 0.720514. Варто використовувати коли найкраща точність має значення.

NMF (негативна матрична факторизація) має середнє значення RMSE приблизно 0.919155 та середнє значення MAE близько 0.720262.

SVD і SVD++ зазвичай використовуються для рекомендаційних систем, де важливі рейтинги користувачів. NMF, натомість, корисний для розкладання матриць з текстовою або зображеннєвою інформацією.

